### DISCUSSION TITLE
swf-version 44 cause my application to lag

### DISCUSSION THREAD

**Discussion by vulkanosaure** [airsdk/Adobe-Runtime-Support#2114] [General]
Hi,
i have an AIR application built with Starling 1.8
because of some ANE dependencies, i recently had to make a big jump from `swf-version` `27 `to `44`, 
and it causes a very noticeable and constant lag when my app is running with `adl `and built with a debug build (when building with release, it's fine). The lag looks like the framerate (normally 60) has dropped to somewhere around 25.

I tested every swf-version, the version that introduced this issue (for me) is `swf-version 38.`, version 37 runs perfectly fine.

My IDE : flash develop, 
air sdk : 33.1.1.929
platform : Windows 12
java version :
```
java version "11.0.14" 2022-01-18 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.14+8-LTS-263)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.14+8-LTS-263, mixed mode)
```

Thanks in advance if you have any inputs

**Comment by ajwfrost** [comment]
Now that's actually fascinating, if I'm interpreting this right: with the same runtime (i.e. you are running within the same AIR SDK), and with a debug build, you see a difference purely based on the swf-version...

We will check this but this could be a massive clue into this issue (see also #85, #307 etc)

thanks

p.s. "platform" says "Windows 12"... I'm hoping 10 or 11 though?! ..

**Comment by vulkanosaure** [comment]
> Now that's actually fascinating, if I'm interpreting this right: with the same runtime (i.e. you are running within the same AIR SDK), and with a debug build, you see a difference purely based on the swf-version...

yeah, in both case, i ran it with air sdk : 33.1.1.929, with the adl command. (on my desktop)

> p.s. "platform" says "Windows 12"... I'm hoping 10 or 11 though?! ..

sorry, windows 11 !

I'm happy if this can connect to an existing problem and add helpful info !

**Reply by ajwfrost** [reply]
Thanks .. although I just had a quick check as I'm keen to try to sort this one out -> but I can't get it to run slowly .. (I'm not using starling, just some other stage3d stuff; have checked I don't have advanced telemetry and I do have a debug build..)

Looking at the code, there's one behavioural change due to a bug in "getBounds" related to 3D objects and scaling, but I don't know that this would cause a performance hit or much of a change. 

... ah ... and of course, the other big change between AIR 26 (SWF 37) and AIR 27 (SWF 38) is that if you use AIR 26 compatibility, then we disable Direct3D 11 and instead use Direct3D 9. Can you check what the Starling information is on the driver details from the trace output it makes during start-up, I think that has the version code in it? Or in Scout? 

Direct3D 11 is starting to cause more problems than it seems to solve... :-(

**Comment by vulkanosaure** [comment]
`[Starling] Display Driver: DirectX9Ex (Standard)`

seems like you got it right. Would that explain it ?

**Reply by ajwfrost** [reply]
Not sure about "explain it", but at least we can add it to the evidence... so it happens (a) if we're using Direct3D 11, (b) if we're in debug mode, (c) if we don't have advanced telemetry set up. It all helps narrow down our investigations..!

**Comment by vulkanosaure** [comment]
more specific info :

output with swf-version 38 :
`[Starling] Display Driver: DirectX11 (Standard Extended)`


output with swf version 37
`[Starling] Display Driver: DirectX9Ex (Standard Extended)`

**Comment by vulkanosaure** [comment]
so if it's something specific to starling 1.8, i don't expect it to be fixed, i'm gonna find a way around it (have a script that switch between swf-version depending on my environment)

**Reply by ajwfrost** [reply]
It's not specific to Starling at all, more about the underlying AIR runtime and how it's interacting with Direct3D. But if you want a simpler workaround you could just force it to use Direct3D 9 the whole time (or just for debug, if you can switch things a little?) - there's a new application descriptor value, `<maxD3D>` which you can set to `9` in order to ensure you use Direct3D 9 even with the latest SWF version.

So e.g. in your app descriptor file before the end `</application>` tag, just paste
```
    <windows>
        <maxD3D>9</maxD3D>
    </windows>
```

thanks

**Reply by vulkanosaure** [reply]
ok thanks, that's a better workaround than what i was gonna do.
Good luck with the investigation

**Reply by vulkanosaure** [reply]
and i can confirm that it solves the lag
